src/egl-client-dmabuf-pool.cpp

```diff
@@ -0,0 +1,317 @@
+                attributesCount += planeAttributes.size();
+            }
+
+            if (bufferData.numPlanes >= 2) {
+                uint32_t modifier_hi = bufferData.modifiers[1] >> 32;
+                uint32_t modifier_lo = bufferData.modifiers[1] & 0xFFFFFFFF;
+                std::array<EGLAttrib, 10> planeAttributes = {
+                        EGL_DMA_BUF_PLANE1_FD_EXT, bufferData.fds[1],
+                        EGL_DMA_BUF_PLANE1_PITCH_EXT, bufferData.strides[1],
+                        EGL_DMA_BUF_PLANE1_OFFSET_EXT, bufferData.offsets[1],
+                        EGL_DMA_BUF_PLANE1_MODIFIER_HI_EXT, modifier_hi,
+                        EGL_DMA_BUF_PLANE1_MODIFIER_LO_EXT, modifier_lo,
+                    };
+
+                std::copy(planeAttributes.begin(), planeAttributes.end(),
+                    std::next(attributes.begin(), attributesCount));
+                attributesCount += planeAttributes.size();
+            }
+
+            if (bufferData.numPlanes >= 3) {
+                uint32_t modifier_hi = bufferData.modifiers[2] >> 32;
+                uint32_t modifier_lo = bufferData.modifiers[2] & 0xFFFFFFFF;
+                std::array<EGLAttrib, 10> planeAttributes = {
+                        EGL_DMA_BUF_PLANE2_FD_EXT, bufferData.fds[2],
+                        EGL_DMA_BUF_PLANE2_PITCH_EXT, bufferData.strides[2],
+                        EGL_DMA_BUF_PLANE2_OFFSET_EXT, bufferData.offsets[2],
+                        EGL_DMA_BUF_PLANE2_MODIFIER_HI_EXT, modifier_hi,
+                        EGL_DMA_BUF_PLANE2_MODIFIER_LO_EXT, modifier_lo,
+                    };
+
+                std::copy(planeAttributes.begin(), planeAttributes.end(),
+                    std::next(attributes.begin(), attributesCount));
+                attributesCount += planeAttributes.size();
+            }
+
+            if (bufferData.numPlanes >= 4) {
+                uint32_t modifier_hi = bufferData.modifiers[3] >> 32;
+                uint32_t modifier_lo = bufferData.modifiers[3] & 0xFFFFFFFF;
+                std::array<EGLAttrib, 10> planeAttributes = {
+                        EGL_DMA_BUF_PLANE3_FD_EXT, bufferData.fds[3],
+                        EGL_DMA_BUF_PLANE3_PITCH_EXT, bufferData.strides[3],
+                        EGL_DMA_BUF_PLANE3_OFFSET_EXT, bufferData.offsets[3],
+                        EGL_DMA_BUF_PLANE3_MODIFIER_HI_EXT, modifier_hi,
+                        EGL_DMA_BUF_PLANE3_MODIFIER_LO_EXT, modifier_lo,
+                    };
+
+                std::copy(planeAttributes.begin(), planeAttributes.end(),
+                    std::next(attributes.begin(), attributesCount));
+                attributesCount += planeAttributes.size();
+            }
+            attributes[attributesCount++] = EGL_NONE;
+        }
+
+        buffer->egl.image = m_renderer.createImageKHR(eglGetCurrentDisplay(),
+            EGL_NO_CONTEXT, EGL_LINUX_DMA_BUF_EXT, nullptr, attributes.data());
```

Comment:
I am missing here checking for errors returned by `eglCreateImage[KHR]`. While I know that it is hardly expected to fail because the parameters passed are under our control here and expected to be valid, it's reasonable to think that at least `EGL_BAD_ALLOC` can be produced (when out of memory)—and it that case it would be nice to at least have something logged out.

The same applies to some of the rest of the GL code around.

Reply:
Returned EGLImageKHR deserves a null-check.

Best I can do for GL code is framebuffer-completeness check at the end of this function.